File: src/c4/inter/For.java
package inter;

import treewalker.TreeWalker;

public class For extends Stmt {

	/*
	 * For ist eine Unterklasse von Stmt. In der Instanzenvariable init_ass wird
	 * die Initialisierungs-Anweisung, in expr die Abbruchbedingung, in iter_ass
	 * die Iterationsanweisung und in stmt der Rumpf abgelegt.
	 */

	Expr expr;
	Assignment init_ass;
	Assignment iter_ass;
	Stmt stmt;

	public For(Assignment a1, Expr x, Assignment a2, Stmt s) {
		expr = x;
		stmt = s;
		init_ass = a1;
		iter_ass = a2;
	}

	public Expr getExpr() {
		return expr;
	}

	public void setExpr(Expr expr) {
		this.expr = expr;
	}

	public Assignment getInit_ass() {
		return init_ass;
	}

	public Assignment getIter_ass() {
		return iter_ass;
	}

	public Stmt getStmt() {
		return stmt;
	}

	public <R, P> R walk(TreeWalker<R, P> walker, P arg) {
		return walker.walkForNode(this, arg);
	}

}

File: src/c4/inter/For.rs
use super::{Assignment, Expr, Node, NodeData, StatementNode, Stmt, StmtData, TreeWalker, Walk};

#[derive(Clone, Debug, PartialEq)]
pub struct For {
    data: StmtData,
    expr: Expr,
    init_ass: Assignment,
    iter_ass: Assignment,
    stmt: Stmt,
}

impl For {
    pub fn new(init_ass: Assignment, expr: Expr, iter_ass: Assignment, stmt: Stmt) -> Self {
        Self {
            data: StmtData::new(),
            expr,
            init_ass,
            iter_ass,
            stmt,
        }
    }

    pub fn get_expr(&self) -> &Expr {
        &self.expr
    }
    pub fn set_expr(&mut self, expr: Expr) {
        self.expr = expr;
    }
    pub fn get_init_ass(&self) -> &Assignment {
        &self.init_ass
    }
    pub fn get_iter_ass(&self) -> &Assignment {
        &self.iter_ass
    }
    pub fn get_stmt(&self) -> &Stmt {
        &self.stmt
    }
}

impl Node for For {
    fn node_data(&self) -> &NodeData {
        &self.data.node
    }
}

impl StatementNode for For {
    fn stmt_data(&self) -> &StmtData {
        &self.data
    }
    fn stmt_data_mut(&mut self) -> &mut StmtData {
        &mut self.data
    }
}

impl<P, R> Walk<P, R> for For {
    fn walk<W: TreeWalker<P, R> + ?Sized>(&self, walker: &mut W, arg: P) -> R {
        walker.walk_for_node(self, arg)
    }
}
